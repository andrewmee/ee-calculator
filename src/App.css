.calculator {
  background: #1795d4;
  border: 1px solid #333;
  border-radius: 20px;
  padding: 20px;
  margin: auto;
  max-width: 400px;
}

.calculator__screen {
  border: 0;
  border-radius: 0.25em;
  box-sizing: border-box;
  font-size: 2rem;
  margin: 2.5%;
  padding: 0.5em;
  text-align: right;
  width: 95%;
}

.calculator__logo {
  display: inline-block;
  height: 60px;
  margin: 2.5%;
  width: 30%;
}

.calculator__button {
  border: 0;
  border-radius: 0.25em;
  color: #555;
  font-size: 2rem;
  line-height: 2em;
  margin: 2.5%;
  padding: 0;
  width: 20%;
}

.calculator__button:hover {
  box-shadow: 0px 0px 10px rgba(0,0,0,0.5);
  color: #000;
}

.calculator__button abbr {
  text-decoration: none;
}

.calculator__button--operation {
  background-color: #ddd;
}

.calculator__button--clear {
  background-color: #fed800;
  float: right;
  vertical-align: bottom;
}

.calculator__inputs {
  clear: right;
  border: 0;
  margin: 10px 0 0;
  padding: 0;
}
