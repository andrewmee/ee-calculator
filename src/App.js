import React, { Component } from 'react';
import logo from './logo.svg';
import Button from './Button';
import './App.css';

class App extends Component {
  constructor() {
    super()
    this.state = {
      currentValue: null,
      storedValue: null,
      operation: null,
    };
    this.handleClick = this.handleClick.bind(this);
    this.calculate = this.calculate.bind(this);
  }

  calculate() {
    const { currentValue, storedValue, operation } = this.state;
    let result = '';

    switch (operation) {
      case 'plus':
        result = parseFloat(currentValue) + parseFloat(storedValue);
        break;
      default:
        result = currentValue;
    }

    this.setState({
      currentValue: result,
      storedValue: null,
      operation: null,
    })
  }

  handleClick(value) {
    const currentValue = this.state.currentValue || '';
    switch (value) {
      case 'clear':
        this.setState({
          currentValue: null,
          storedValue: null,
          operation: null,
        });
        break;
      case 'equals':
        this.calculate();
        break;
      case 'plus':
        // TODO: if operation already stored, trigger a calculation
        this.setState({
          currentValue: '',
          storedValue: currentValue,
          operation: value,
        });
        break;
      case '.':
        if (currentValue.indexOf('.') >= 0) {
          break;
        }
      default:
        this.setState({
          currentValue: `${currentValue}${value}`
        });
    }
  }

  render() {
    return (
      <form className="calculator">
        <input className="calculator__screen" readOnly value={this.state.currentValue || 0} />
        <img src={logo} className="calculator__logo" alt="Equal Experts" />
        <Button onClick={this.handleClick} value="clear" type="operation" label={
            <abbr title="Clear">C</abbr>
          } />
        <fieldset className="calculator__inputs">
          <Button onClick={this.handleClick} value="8" />
          <Button onClick={this.handleClick} value="7" />
          <Button onClick={this.handleClick} value="9" />
          <Button disabled value="divide" type="operation" label={
            <abbr title="Divide">&divide;</abbr>
          } />
          <Button onClick={this.handleClick} value="4" />
          <Button onClick={this.handleClick} value="5" />
          <Button onClick={this.handleClick} value="6" />
          <Button disabled value="times" type="operation" label={
            <abbr title="Times">&times;</abbr>
          } />
          <Button onClick={this.handleClick} value="1" />
          <Button onClick={this.handleClick} value="2" />
          <Button onClick={this.handleClick} value="3" />
          <Button disabled value="minus" type="operation" label={
            <abbr title="Minus">&minus;</abbr>
          } />
          <Button onClick={this.handleClick} value="0" />
          <Button onClick={this.handleClick} value="." />
          <Button onClick={this.handleClick} value="equals" label="=" />
          <Button onClick={this.handleClick} value="plus" type="operation" label={
            <abbr title="Plus">+</abbr>
          } />
        </fieldset>
      </form>
    );
  }
}

export default App;
