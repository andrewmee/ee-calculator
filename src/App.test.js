import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

it('renders correctly', () => {
  const wrapper = shallow(<App />);
  expect(wrapper).toMatchSnapshot();
});

describe('calculate', () => {
  it('works with no operation', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({
      currentValue: '3',
      operation: null,
    });
    wrapper.instance().calculate();
    expect(wrapper.state('currentValue')).toBe('3');
  });

  it('handles simple plus', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({
      currentValue: '3',
      storedValue: '3',
      operation: 'plus',
    });
    wrapper.instance().calculate();
    expect(wrapper.state('currentValue')).toBe(6);
  });

  it('handles decimal plus', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({
      currentValue: '3.5',
      storedValue: '3',
      operation: 'plus',
    });
    wrapper.instance().calculate();
    expect(wrapper.state('currentValue')).toBe(6.5);
  });
})

describe('handleClick', () => {
  it('stores current value', () => {
    const wrapper = shallow(<App />);
    wrapper.instance().handleClick('9');
    expect(wrapper.state('currentValue')).toBe('9');
  });
  
  it('clears current value', () => {
    const wrapper = shallow(<App />);
    wrapper.instance().handleClick('9');
    wrapper.instance().handleClick('clear');
    expect(wrapper.state('currentValue')).toBe(null);
  });
  
  it('appends to current value', () => {
    const wrapper = shallow(<App />);
    wrapper.instance().handleClick('9');
    wrapper.instance().handleClick('9');
    expect(wrapper.state('currentValue')).toBe('99');
  });
  
  it('appends decimals to current value (only once)', () => {
    const wrapper = shallow(<App />);
    wrapper.instance().handleClick('9');
    wrapper.instance().handleClick('.');
    wrapper.instance().handleClick('0');
    expect(wrapper.state('currentValue')).toBe('9.0');
    wrapper.instance().handleClick('.');
    expect(wrapper.state('currentValue')).toBe('9.0');
  });

  it('triggers calculation on equal', () => {
    const wrapper = shallow(<App />);
    wrapper.instance().handleClick('9');
    wrapper.instance().handleClick('plus');
    wrapper.instance().handleClick('9');
    wrapper.instance().handleClick('equals');
    expect(wrapper.state('currentValue')).toBe(18);
  });
});
