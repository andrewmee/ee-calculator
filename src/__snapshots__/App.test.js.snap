// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  className="calculator"
>
  <input
    className="calculator__screen"
    readOnly={true}
    value={0}
  />
  <img
    alt="Equal Experts"
    className="calculator__logo"
    src="logo.svg"
  />
  <Button
    label={
      <abbr
        title="Clear"
      >
        C
      </abbr>
    }
    onClick={[Function]}
    type="operation"
    value="clear"
  />
  <fieldset
    className="calculator__inputs"
  >
    <Button
      onClick={[Function]}
      value="8"
    />
    <Button
      onClick={[Function]}
      value="7"
    />
    <Button
      onClick={[Function]}
      value="9"
    />
    <Button
      disabled={true}
      label={
        <abbr
          title="Divide"
        >
          ÷
        </abbr>
      }
      type="operation"
      value="divide"
    />
    <Button
      onClick={[Function]}
      value="4"
    />
    <Button
      onClick={[Function]}
      value="5"
    />
    <Button
      onClick={[Function]}
      value="6"
    />
    <Button
      disabled={true}
      label={
        <abbr
          title="Times"
        >
          ×
        </abbr>
      }
      type="operation"
      value="times"
    />
    <Button
      onClick={[Function]}
      value="1"
    />
    <Button
      onClick={[Function]}
      value="2"
    />
    <Button
      onClick={[Function]}
      value="3"
    />
    <Button
      disabled={true}
      label={
        <abbr
          title="Minus"
        >
          −
        </abbr>
      }
      type="operation"
      value="minus"
    />
    <Button
      onClick={[Function]}
      value="0"
    />
    <Button
      onClick={[Function]}
      value="."
    />
    <Button
      label="="
      onClick={[Function]}
      value="equals"
    />
    <Button
      label={
        <abbr
          title="Plus"
        >
          +
        </abbr>
      }
      onClick={[Function]}
      type="operation"
      value="plus"
    />
  </fieldset>
</form>
`;
